### ISSUE TITLE
multiple independent co-operative core processes

### ISSUE THREAD

**Issue by tvanderstad** [lockbook/lockbook#2207]
experienced by me on windows, natural consequence of no file locks or other inter-process communication

**Comment by Parth** [comment]
I talked with @Kousay-Jebir and we strategized much of the following for him to flesh out and send a proposal for in more detail

# IPC approach

- use some networking technology or other IPC method that behaves like this:
- lb init will attempt to become a leader (by binding to a port or other such mechanism)
- if leader it will expose an interface that followers can talk to, in the manner of all the functions specified on the `lb-rs`: for reference here is a full list: https://docs.rs/lb-rs/latest/lb_rs/struct.Lb.html
- the leader's client can continue to use normal function calls not going through this abstraction
- a follower can "connect" and have core functions served through this IPC interface
- upon a connection failure each follower will try to become a leader, after a short while they will try to connect to a new leader
- leader's experience should not be diminished in any observable manner, max observable acceptable is the addition of a single branch for every call (if statement, enum match, dynamic dispatch, static dispatch, etc).

# enabled use cases
- cli can talk to desktop which is great for users (less data replicated), great for developers (because of massively enhanced debugability), and it's great for our server (because of significantly less data being synced needlessly).
- multiple windows support unlocked for situations where a parent process cannot be easily shared
- core-like interface also enables someone to use their desktop as a proxy for their phone (the travis bad network use case).

# some areas of uncertainty
- I reckon that we'll be creating a `lb-like` trait and implementing it either directly (for the leader) with the existing functionality or "indirectly" by using this IPC.
- should this trait expose the async versions of the functions or the blocking versions of these functions
- I have the feeling that since lb is async and the tech chosen is probably going to be thinking in async terms that this trait will be async, although async trait support in rust is nascent and there could be some struggle there. I'm open to basically wrapping some significant amount of our logic in an enum with multiple implementations.
- I think this technology should probably support concurrent access to core, I think there could be situations where we have to use a single pipe or channel or stream of some sort and correlate requests and responses. Maybe the latency is low enough of all core calls where we could simply block each request waiting for a response, but I think that would limit the flexibility of the API. Things like `get_usage` are up to the network how long they take and it would be sad if this API was blocked on them. I think however these apis may be fast enough for each request to form it's own connection to simplify the code, but we should test to find out. The fact that a lot of IPC frameworks (at-least in the crate `interprocess`) say they "bypass the network stack implies to me that this may truly be fast enough, but research and testing will tell us more.
- although these low overhead things may preclude the travis offline use-case and so maybe more programmer sophistication is required regardless.
- something like tonic could be viable, but not sure the complexity is worth the hassle

**Comment by Kousay-Jebir** [comment]
**I talked with @Parth  yesterday about a possibly different solution**

- **Initial Proposal – Memory-Sharing Model**  
  I suggested using a `memmap2` file to hold a single serialized `Lb` instance in shared memory. Every process would treat that shared region as the “one true” state and acquire an in-memory lock before reading or writing.

- **Motivation: Leaderless Peer Model**  
  Rather than designating one process as the “leader” and having others proxy through it, the goal was to let each process be equal simply racing to access the same shared state.

- **Prototype & Proof-of-Concept**  
  I built a pseudo-code ish small prototype mapping a file, and using a lock to allow only one process to instantiate Lb at once and put it in the file 
- **Fundamental Blocker**  
  The real `Lb` struct holds live resources—open DB connections, network sockets, Tokio tasks, `Arc` internals, file handles, etc.—which apparently cannot be serialized into shared memory and re hydrated in another process.

- **Inevitable Leader/Follower Model**  
  This limitation means one process must truly “own” the `Lb` instance. Other processes will need to **delegate** their calls to that owner.

- **Delegation Mechanisms Under Consideration**  
  1. **Shared-Memory RPC**  
     - Followers write requests into a memory-mapped buffer; the leader reads, executes, and writes responses back.  
  2. **Conventional IPC**  
     - Followers open a Unix socket or named pipe and send/receive framed messages.  
- **Preliminary Conclusion**  
  While a shared-memory RPC is intellectually appealing, its complexity and synchronization costs may exceed any performance gain over a straightforward IPC socket approach.

**Comment by Parth** [comment]
Yeah that sounds good, sound like we'll be going for a traditional RPC style interaction for the follower processes, probably over the network to enable @tvanderstad's use cases

**Comment by tvanderstad** [comment]
In this design would I be able to use my phone using my desktop as a proxy server without them actively connected? I imagined my use cases would only be met by a design where a client exposes a server-like API though I'm not sure how that meets the other needs here

**Comment by Kousay-Jebir** [comment]
> In this design would I be able to use my phone using my desktop as a proxy server without them actively connected? I imagined my use cases would only be met by a design where a client exposes a server-like API though I'm not sure how that meets the other needs here
### I think it does not, here's my thought about why
1. **Fallback‑to‑Leader Rule**  
   - Every process (desktop windows, CLI, phone) first tries to connect to the existing leader.  
   - If it fails, it _becomes_ the leader and spins up its own `Lb` core in its own “universe.”  

2. **Classic Same‑Device Scenarios**  
   - On your desktop, multiple windows or a CLI shell work perfectly: they delegate to the running desktop app’s core, and only fall back if you explicitly close that app.  

3. **Phone‑Desktop Proxy Breaks Under Fallback**  
   - When your phone can’t reach the desktop, it simply becomes leader and runs its own core.  
   - Later, when the desktop reconnects, it too spins up a core—so you end up with two independent instances instead of a true proxy.

In this design the leader is indeed exposing a server-like API.

@tvanderstad  are you suggesting that even if your phone can't reach the desktop, it should still persist requests for the desktop when it becomes online again or are you if your phone can still just do his work on his own when he's not connected to the desktop (**in this case yes it can**)

**Comment by Parth** [comment]
Okay let me clarify some things. 

- We **are** going to go for a a traditional tcp connection and perform RPC over that. By default it'll try to connect to a predefined port on loopback and if it succeeds it will become a server for other clients. The leader will communicate over the lower latency direct function calls as usual, but other clients, instead of being blocked indefinitely will be able to communicate with this leader over tcp.
- I was curious about lower latency approaches so I had Kousay read about these things and what their implications could be, but because that will never work across machines they're out.
- I was curious about even lower latency approaches like shared memory (no serialization?) but due to their complexity and lack of interop with rust primitives (like all the dynamically allocated structs) they're out.

The traditional RPC will meet the use cases:
- remote proxy (with some config for an advanced user to specify a desktop address and a fallback).
- enhanced debug ability
- reduced device footprint

@Kousay-Jebir you should try to make it more clear what things are coming from ai vs. yourself. Just quote and say what chatgpt said or share a link to the conversation directly, it provides more clarity. Additionally, obviously, feel free to say if you don't know the answer to something, that's fine too.

**Comment by Parth** [comment]
@Kousay-Jebir wanted some clarity on how this may help someone with a bad internet connection:

if someone has a bad internet connection this would allow their devices to be in sync with each other, and then can off load broader network reconciliation to a machine more suited to that task (a desktop that stays connected that can trickle the bytes to the server overnight).

I think the first stage of this will be local only, but a sophisticated user will be trivially able to bind their desktop app to a port their phone can access and build core to also connect to this port on their desktop, enabling this use case.

Much later we may be able to negotiate this automatically for people so their phone will sync with their desktop rather than our server if the desktop has all the information they could want.

**Comment by tvanderstad** [comment]
I kind of think the proxy server business should be treated as a separate problem. Like you say, I'd have to build core to connect to my desktop, and the app would be unusable unless connected to my desktop. I might use this on my tablet which I generally leave in my home, but my phone I want to be out-and-about with, so I think I'd need to be able to sync changes to whichever 'server' is accessible at the time. Like, my phone should still have a local copy of my lockbook that's usable when it's disconnected from my desktop, and should be able to sync its changes to either place.

**Comment by tvanderstad** [comment]
In this situation my desktop acts as the server, offering an API for making atomic file system changes, delegating situations requiring conflict resolution to my phone, and receiving new file contents. The effects of those changes in turn get stored as local changes that it relays to the server when my home internet connection is restored.

**Comment by Parth** [comment]
I was thinking you'd configure your phone to connect to your computer, it would fall back locally if unsuccessful. While connected to your computer it would sync it's local core in the background.

giving you access to both, basically.

**Comment by Parth** [comment]
I think most of  the programming here is going to be similar, it'll be something that looks like a client - server. The questions is what technology do we reach for. TCP or something lower level, platform specific, that doesn't involve the network stack.

We've covered the intra process stuff pretty well. But both inter-process, and inter-host communication will program very similarly.

Given our main client will just be doing function calls, reaching for network stuff seems fine for the CLI.

Then with some experimental code if we can fallback and negotiate changes among two cores trivially then I think we get access to some cool stuff long term. I think it's okay if that never materializes and the CLI has a slightly slower connection locally.

**Comment by Parth** [comment]
I don't think you're skeptical of the fallback you're likely skeptical off the core - core reconciliation. I think it could be light. I don't think it's essential but I think the possibility of us doing something like this at scale makes me want to reach for something like TCP now instead of the other things in the `interprocess` crate.

we could just not, and then when you fallback to your original core you'd sync with the server. maybe you don't like that, you made changes on a device and you want to see them on this device. If the proxy url is exposed in settings I think a customer who is savy enough to do this will understand what's going on. let's pretend for a second this isn't acceptable:

your local core, in the background will attempt to safe-write all it's changes, possibly stage meta changes and if validate works, apply them. if successful do a (fast & atomic) byte by byte copy of that core and overwrite yours. Now you're up to date as you leave the house.

alternatively the proxy server you're describing could live in core itself. there are only three endpoints we'd need to proxy follower core tries to achieve a clean sync through the api in the background, can be a long running process. it'll negotiate it's own updates with the host server gradually. When this succeeds, either core that's more ahead can assimilate the other core if desired.

I think the initial implementation is not going to do any of that and will just expose access to a single core to multiple entities and only during time of contention, with the leader ending up as the longest running client. I'm just outlining what a network based approach could enable.

**Comment by tvanderstad** [comment]
I'm aiming to be able to experience Lockbook's file sync on my local network without my local network being connected to the internet, and yeah, it's the core-core reconciliation I'm concerned about.

I think core-core reconciliation should happen by core exposing the server API. Then one core just syncs using the other core as a server. It doesn't need to be a clean sync. It isn't more of a gradual or long-running process than a regular sync. A regular sync already assimilates two cores. I think the client core inherits the last synced time of the core acting as a server. I think the one with the more recent last synced time should act as the server, otherwise the client core will push server changes to the server core, and if there are conflicts they will be resolved as if the server changes are local changes which might give priority to the wrong branch. I don't think it's a big deal either way though.

I wouldn't want to byte-by-byte copy a core over because my desktop may have it's own local changes that I made there or that it aggregated from other devices on my local network. 

Connecting my phone to my desktop to view the files on my desktop's core through my phone feels cool but isn't really the same thing. Multi-window & debug wins etc are noted. I agree with TCP as the more flexible transport technology but doesn't feel required if we regard the core-core sync/proxy server situation as a later problem that we're down to solve in a different way.

**Comment by Parth** [comment]
Ah you would want core to not proxy those endpoints to the production server but use those endpoints *as* the method of reconciliation between the two cores.

I guess this goes back to a similar back and forth we have, to me sync doesn't feel like the right tool here because either party can decrypt the content, it doesn't have the same constraints as the server. But yeah down to discuss the implementation details when we're ready.

I think we're on the same page about the fact that it should probably be easy to do this, come from core, and possibly ideally come from something like leaving the desktop app open on some powerful machine.

And then the shape of the core-core recon process is something we can discuss later.

**Comment by Kousay-Jebir** [comment]
> @Kousay-Jebir you should try to make it more clear what things are coming from ai vs. yourself. Just quote

i used a temporary chat so i don't have access to it now, but basically i asked ai to tell me the pros and cons of each approach
and ask it to help me a little with writing comments here so that i can clarify my thought process more.

>I was curious about even lower latency approaches like shared memory (no serialization?) 

Ye so i was curious about wether is it possible to save Lb in memory (thus serializing the struct to persist it) in a way that any process can use it directly as if it was its own (without having to talk to the process who created it ) found out later that it's impossible to achieve this in our case due to the complexity of Lb,

**Comment by Parth** [comment]
use of ai or any tool really is fine, just make it's presence obvious by quoting it or something, that way we know whether we're talking to you or "someone" else.

yeah just to be clear the thing that rules it out isn't Lb's inherent complexity, anything that would use a `String`, `Vec`, or `Mutex` would not be able to use this approach. It's possible we could look into custom allocators and that world of things, but this thread contains a lot of context for why maybe we would want tcp anyways

**Comment by Kousay-Jebir** [comment]
Okay, i've read what @tvanderstad  and @Parth  talked about and here is my summary of everything mentioned based on what i understood:

- we will only  use an RPC over TCP approach for now
- There's going to be some sort of two modes.

**1. Default (no proxy configured):**

Every process starts by attempting to bind to the well‑known loopback port.

- If bind succeeds, this process is the leader (it owns the Lb core).
- If it fails then this process is a follower and it connects to the leader over TCP for all API calls.
- If a follower ever loses connection to the leader, it probably should retry a quick few times before starting some sort of election for a new leader.

**2. Proxy Mode (manually configured by a user who knows what he's doing) :**

- Initially, the process (in this example the phone) first attempts to connect to the configured proxy address instead of trying to bind itself like mentioned above.
- If the proxy connection succeeds, it remains a follower to that proxy leader.
- If it cannot reach the proxy, it simply runs its own local core. I don't know if it should also retry connecting to the proxy automatically in a polling manner for example.

>If it cannot reach the proxy, it simply runs its own local core.

From what i understood you guys mentioned the following problem : how to handle syncing the two cores that we have now

**the solution :**  Rather than copying entire cores byte‑for‑byte, we base it off on whichever core has the more recent `last_synced` to reconcile.
Travis pointed out that this doesn't have to be a clean sync bur rather progressive.

>I guess this goes back to a similar back and forth we have, to me sync doesn't feel like the right tool here 

this solution will be further discussed

As travis mentioned,  desktop may have its own local changes directly byte copying risks overwriting stuff. but syncing as if desktop was truly a server each core’s edits.

for now, you manually  configure for example your desktop as a proxy.

**All of this seems to check all the boxes :** 

- local sync without reaching out to the internet (based on the assumption that one client will be viewed as an actual server) which fixes the bad internet issue
- this clearly works well in the multiple window / cli + egui use case
- one sync over the internet ends up happening

I hope this touches on all the areas in the correct way

**Comment by Parth** [comment]
Yeah, I would emphasize that. The default mode is the main thing we want for now, but that is a good summary of the situation. The other mode we can flesh out and explore later.

But maybe this is enough clarity for you to start trying stuff

**Comment by tvanderstad** [comment]
On the "some sort of election for a new leader front" I would say that whoever has the port is the leader. Leader election is distributed systems is hard because it's in the context of a few servers, each of which could fail at any time, and the system still has to agree on the leader. In practice this would be delegated to a system like [Consul](https://developer.hashicorp.com/consul/docs/automate/application-leader-election) or [Zookeeper](https://zookeeper.apache.org/doc/r3.9.3/recipes.html#sc_leaderElection) (or in Google, [Chubby](https://static.googleusercontent.com/media/research.google.com/en//archive/chubby-osdi06.pdf)) which is deployed to a few computers that run some tricky protocol and the rest of the system just asks them who the leader should be. We can instead delegate to the OS, which will choose who the leader should be by who gets the port. I assume that the OS will free the port immediately if the process that owns it terminates for any reason so the port ownership effectively manages the leader's 'lease'.

**Comment by tvanderstad** [comment]
On "I reckon that we'll be creating a lb-like trait" may I recommend instead considering an enum with one variant for each situation. May help avoid weirdness with async traits and general trait-related confusion.

**Comment by Kousay-Jebir** [comment]
> On "I reckon that we'll be creating a lb-like trait" may I recommend instead considering an enum with one variant for each situation. May help avoid weirdness with async traits and general trait-related confusion.

Yep that is the idea we were heavily leaning towards during our first talk

**Comment by Parth** [comment]
> I have the feeling that since lb is async and the tech chosen is probably going to be thinking in async terms that this trait will be async, although async trait support in rust is nascent and there could be some struggle there. I'm open to basically wrapping some significant amount of our logic in an enum with multiple implementations.


Down to do the enum thing, I think we should try the trait and see how it goes to gather more info as well tho

**Comment by Kousay-Jebir** [comment]
I started writing some very high level ideas just to experience the two approaches , i tried my best to keep client code untouched for whoever implements lb-rs.

In the enum approach i like how i didn't have to use any crates but i don't like it how i needed to write 2 implementations for the methods (although one of the implementations is just repetitive code and could probably be automated using some macro ).

In the trait approach i probably didn't come out with the most elegant solution as i am quite new with the struggles of traits and trait objects.


**checkout bin folder** 
[Link to my playground repo](https://github.com/Kousay-Jebir/lb-rs-prototype)

My plan is to move to start experimenting with the actual Network level logic if this gets approved as some primary proof of concept.

**Comment by Parth** [comment]
- this is good stuff 
- yeah the main thing that requires `async_trait` (note to myself for the future) is the fact that `init`'s signature needs to return `&dyn Lb`. I think requiring `async_trait` (the crate) rules out the trait approach in my eyes
- I think it's time to put this in a PR, with any of these approaches there will be a decent amount of boiler plate stuff (like you said @Kousay-Jebir).
- As a next step I would try to just solve this problem for the functions in `account_service` and `init` and we can start to dive into some of the details of organizing the code, the files, the modules, etc.

**Comment by Parth** [comment]
@Kousay-Jebir I just realized -- I can make `init` not-async. Which means we can probably drop the requirement for `async_trait`, probably making the async approach viable or at-least worth considering for the `service/account.rs` proof of concept. Maybe you could do two PRs to really evaluate each approach properly. I think the details here are going to push us in one direction or another. For example the trait implementation must live in a single file, but with the enum we don't have that requirement. But we may not need that requirement if all the boiler plate for either approach is limited to a single area.

I can see us doing quite a bit of iteration on each approach in a PR then closing the one we don't have conviction in and continuing on with the one that seems the most promising for the rest of the feature.

**Comment by Kousay-Jebir** [comment]
Sounds good. I don't understand this 
> the trait implementation must live in a single file, but with the enum we don't have that requirement.

**Comment by Kousay-Jebir** [comment]
I've been thinking on how would i send errors returned by Lb's api in `LbResult` over TCP
I noticed that `LbErr` can't be serialized and sent over the network due to its `backtrace` field. So I've been thinking on how can i send something over the network that would explain what kind of error we have and be able to reconstruct it later after reading the TCP response.
Example : 
in `account.rs`  we are potentially returning different kinds of `LbErrKind`. My problem is how can i know upon reading the TCP response what is that error.
the thought that came to mind is some sort of function that could return the exact type from and to a string of its name 
example : if the error is `AccountExists` we send the string `AccountExists` and in the rpc module you would call a function on that string that returns the type `AccountExists` and for the backtrace, just keep it to `None` all the time and for the backtrace field no need to trasmit it just keep it to None all the time for followers since we don't need it replicated in multiple processes

**Comment by Parth** [comment]
> Sounds good. I don't understand this
> 
> > the trait implementation must live in a single file, but with the enum we don't have that requirement.

When you `impl` something on a struct, you can do so multiple times, as long as the function signatures aren't duplicated. See our egui client, we have impls on the same scruct scattered across different files.

However when you `impl Trait for Struct` that whole `Trait`'s definition must live in a single file. 

So with a `Trait` we're likely to have a file that looks something like this: https://github.com/lockbook/lockbook/blob/master/libs/lb/lb-rs/src/blocking.rs

This is a file with every function specified in one spot. This may be desirable in some situations, like ideally the server / client business will be restrained to one file. But it may be undesirable if **all** the current functions organized in the `service/` directory needed to be in a single file.

**Comment by Parth** [comment]
On the `LbErr` front we can just `Serialize` backtrace our selves by `impl Serialize for LbErr` and using a debug implementation or something similar if the backtrace is present. Can do something similar for `Deserialize`. 

Here is an example of us doing this: https://github.com/lockbook/lockbook/blob/f434f74f0d428e3e7b50996beddcdbcc34ed00e7/libs/lb/lb-rs/src/model/errors.rs#L248

**Comment by Parth** [comment]
This sort of discussion may be better off as comments on a draft PR as we're no longer exploring the requirements of completing this feature but rather specific implementation details.

**Comment by Kousay-Jebir** [comment]
First pr for the enum implementation 
https://github.com/lockbook/lockbook/pull/3590

**Comment by Parth** [comment]
dupe of [#2152](https://github.com/lockbook/lockbook/issues/2152)
